Lib/unittest/mock.py

```diff
@@ -2758,13 +2758,15 @@ def create_autospec(spec, spec_set=False, instance=False, _parent=None,
                                f'[object={spec!r}]')
     is_async_func = _is_async_func(spec)
 
-    entries = [(entry, _missing) for entry in dir(spec)]
+    base_entries = {entry: _missing for entry in dir(spec)}
     if is_type and instance and is_dataclass(spec):
         dataclass_fields = fields(spec)
-        entries.extend((f.name, f.type) for f in dataclass_fields)
+        entries = {f.name: f.type for f in dataclass_fields}
```

Comment:
How come we don't end up using `UnionType` here from the `int | None` annotation?

Reply:
Because I override `entries` with `default_entries` here: https://github.com/python/cpython/pull/124724/files#diff-347d0254250a1ab7ab8e31b405e2c35b74cd2838df4ee74f1b658a459eb91f1aR2765

This way `field(default=0)` has a priority over the annotation.

Reply:
Two questions from that:

 1. What happens with a class such as this?
 
    ```python
    @dataclass
    class SampleClass:
        sample_attr: int | None
        
    ```

2. What does the mock machinery do with the `field` object it gets in the case you show above?

Reply:
1. In this case `spec` would be `UnionType`, because we have no default
2. Mock does nothing special to `field`, `dataclass` itself sets `SampleClass.sample_attr = default`, and we just use `getattr` to get the existing default

Reply:
Would it be possible to add a unit test (or more if necessary) showing this?

Reply:
Done!